sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -636,6 +641,102 @@ object GpuCast extends Arm {
+        // cast each element to string
+        withResource(
+          doCast(child, elementType, StringType, ansiMode,
+                 legacyCastToString, stringToDateAnsiModeEnabled)
+        ) { strChildWithNull =>
+          withResource(strChildWithNull.replaceNulls(nullRep)) { strChild =>
+            if (legacyCastToString) {// add a space string to each non-null element
+              withResource(ColumnVector.fromScalar(space, child.getRowCount.toInt)) { spaceVec =>
+                withResource(
+                  ColumnVector.stringConcatenate(Array(spaceVec, strChild))
+                ) { AddSpace =>
+                  withResource(child.isNotNull) {_.ifElse(AddSpace, strChild)}
+                }
+              }
+            }
+            else { strChild.incRefCount }
+          }
+        }
+      }
+
+      /**
+       * If the first char of a string is ' ', remove it
+       * @param strVec a string type column vector
+       */
+      def removeFirstSpace(strVec: ColumnVector): ColumnVector = {
+        if (legacyCastToString){
+          withResource(ColumnVector.fromScalar(space, strVec.getRowCount.toInt)){spaceVec =>
+            withResource(strVec.substring(0,1)) { fstChar =>
+              withResource(strVec.substring(1)) { remain =>
+                withResource(fstChar.equalTo(spaceVec)) {_.ifElse(remain, strVec)}
+              }
+            }
+          }
+        }
+        else {strVec.incRefCount}
+      }
+
+      /**
+       * Add brackets to each string. Ex: ["1, 2, 3", "4, 5"] => ["[1, 2, 3]", "[4, 5]"]
+       * @param strVec a string vector
+       */
+      def addBrackets(strVec: ColumnVector): ColumnVector = {
+        withResource(
+          Seq(left, right).safeMap(s => ColumnVector.fromScalar(s, numRows))
+        ) { case Seq(leftColumn, rightColumn) =>
+          withResource(ArrayBuffer.empty[ColumnVector]) { columns =>
+            columns += leftColumn.incRefCount
+            columns += strVec.incRefCount
+            columns += rightColumn.incRefCount
+
+            ColumnVector.stringConcatenate(empty, nullRep, columns.toArray)
+          }
+        }
+      }
+      /* -------------------------------- helper functions -----------------------*/
+      // cast child column to string type
+      withResource(input.getChildColumnView(0)) { childView =>
+        withResource(castChildToStr(childView)){ stringChild =>
+          withResource(input.replaceListChild(stringChild)) {pureStrArr =>
+            // concatenate each row. cast from list column to string column
+            withResource(
+              pureStrArr.stringConcatenateListElements(sep, nullRep, true, true)
+            ) { withoutBrackets =>
+              withResource(removeFirstSpace(withoutBrackets)){withoutBrackets =>
+                // add "[ ]"
+                withResource(addBrackets(withoutBrackets))(
+                  _.mergeAndSetValidity(BinaryOp.BITWISE_AND, pureStrArr)
```

Comment:
I don't think we need to merge validity here, because `stringConcatenateListElements` already achieved it.